### PR TITLE
Naive fix for #13171

### PR DESCRIPTION
**Pull request by Sacha0**
Ref JuliaLang/LinearAlgebra.jl#257. This fix's simplicity has me convinced that I'm either missing something or made a mistake.

### CODE REVIEW

File: base/operators.jl [hunk 1]
```diff
@@ -293,7 +293,7 @@ fldmod1{T<:Real}(x::T, y::T) = (fld1(x,y), mod1(x,y))
 fldmod1{T<:Integer}(x::T, y::T) = (fld1(x,y), mod1(x,y))
 
 # transpose
-transpose(x) = x
+transpose(x) = error("transpose not implemented for $(typeof(x)). Consider permutedims.")
```
Comment (kshyatt) [inline base/operators.jl]: Why is this an `error` and not something like `ArgumentError`? I thought we were trying to move away from `error()`?

Comment (Sacha0) [inline base/operators.jl]: No reason beyond mimicking https://github.com/JuliaLang/julia/blob/9a66205d071bd013b5c0c607764a0885dd4e38f3/base/abstractarraymath.jl#L9-L10. Should I change it to `ArgumentError`, and if so should I change the linked definitions to `ArgumentError`s as well? Thanks for reviewing! :)

Comment (kshyatt) [inline base/operators.jl]: I think we should keep this PR self-contained for now - you can always open a PR for those others later. But yeah, I think something like `ArgumentError` is more useful than `error()`.

Comment (Sacha0) [inline base/operators.jl]: Cheers, revised to `throw` an `ArgumentError`. Thanks!

**Comment by ivarne** [comment]
Ideally it should be a ``MethodError`` (right?) But as long as we don't have https://github.com/JuliaLang/julia/issues/7512, we can't print the message in `MethodError`.

**Comment by andreasnoack** [comment]
I think packages will hit this more often. @JeffBezanson was also not happy about adding a lot of no-op `(c)transpose` methods including the cases `Char` and `AbstractString`. The question is therefore if we should go all the way and declare `transpose` separate from `permutedims` and require that arrays of strings or chars use `permutedims`.

As a follow up on the last comment, I'll mention that I'm trying to rebase my transpose immutable branch and realized that the `Transpose` immutable can't be restricted to wrap `AbstractMatrix` and can't be a subtype of `AbstractArray` either since people will definately want to do `A'x` for their custom linear operators that are not subtypes of `AbstractMatrix`. In that sense, we are also on the path to separating `transpose` from `permutedims` where the latter only deals with `AbstractArrays`.

**Comment by Sacha0** [comment]
> I think packages will hit this more often. @JeffBezanson was also not happy about adding a lot of no-op (c)transpose methods including the cases Char and AbstractString. The question is therefore if we should go all the way and declare transpose separate from permutedims and require that arrays of strings or chars use permutedims.
> 
> As a follow up on the last comment, I'll mention that I'm trying to rebase my transpose immutable branch and realized that the Transpose immutable can't be restricted to wrap AbstractMatrix and can't be a subtype of AbstractArray either since people will definately want to do A'x for their custom linear operators that are not subtypes of AbstractMatrix. In that sense, we are also on the path to separating transpose from permutedims where the latter only deals with AbstractArrays.

+1 to decoupling speciously coupled concepts!

**Comment by andreasnoack** [comment]
Let's then try to remove the definitions for `AbstractString` and `Char` and adjust the tests.

**Comment by Sacha0** [comment]
> Let's then try to remove the definitions for AbstractString and Char and adjust the tests.

Done. Thoughts? Thanks and best!

**Comment by andreasnoack** [comment]
Looks like `datafmt.jl` also needs some adjustments.

**Comment by Sacha0** [comment]
> Looks like datafmt.jl also needs some adjustments.

Points for CI. What say you now, Travis?

**Comment by tkelman** [comment]
This needs to be a deprecation, not an immediate error.

**Comment by Sacha0** [comment]
Thanks for the review and merge!

**Pull request merged**